### ISSUE TITLE
Model structure issue

### ISSUE THREAD

**Issue by AKun88888888** [taco-group/DecAlign#11]
Thank you for your work. I have the following questions:
1.The results obtained using the MOSI dataset and default parameters exhibit significant discrepancies. Could it be due to my configuration?
2. In the models network structure, I found that the line `final_rep = torch.cat([fusion_rep_trans, fusion_rep_cma], dim=1) # [N, 6*d_l]` does not seem to have information about homogeneous features, which I don't quite understand

**Comment by Raymond-Qiancx** [comment]
Thank you for your interest in our work and for raising these questions!

**Regarding Question 1 (Result discrepancies on MOSI)**

The result variations might be caused by several factors:
- **Random seed**: Please ensure consistent random seed settings.
- **BERT model version**: We use `bert-base-uncased`. Different versions or download sources may cause slight variations.
- **Data preprocessing**: Please verify your `mosi_data.pkl` is preprocessed consistently with ours.
- **Hardware differences**: Different GPU architectures may introduce minor numerical differences.

Could you share the specific metrics you obtained compared to the expected values? This would help us better diagnose the issue.

**Regarding Question 2 (Homogeneous features not in final_rep)**

Thank you for catching this! This was an oversight when we merged the ablation study versions into the final codebase. The homogeneous features should indeed contribute to the final representation.

We have now fixed this issue. The updated fusion strategy is:

```python
# Homogeneity branch: average pooling over time dimension
c_l_avg = c_l.mean(dim=2)
c_a_avg = c_a.mean(dim=2)
c_v_avg = c_v.mean(dim=2)
fusion_rep_homo = torch.cat([c_l_avg, c_a_avg, c_v_avg], dim=1)  # [N, 3*d_l]

# Fuse heterogeneous paths and concatenate with homogeneous
fusion_rep_hete = fusion_rep_trans + fusion_rep_cma  # [N, 3*d_l]
final_rep = torch.cat([fusion_rep_hete, fusion_rep_homo], dim=1)  # [N, 6*d_l]

Thank you again for your feedback! Please let me know if this addresses all your concerns.

**Comment by AKun88888888** [comment]
> Thank you for your interest in our work and for raising these questions!
> 
> **Regarding Question 1 (Result discrepancies on MOSI)**
> 
> The result variations might be caused by several factors:
> 
> * **Random seed**: Please ensure consistent random seed settings.
> * **BERT model version**: We use `bert-base-uncased`. Different versions or download sources may cause slight variations.
> * **Data preprocessing**: Please verify your `mosi_data.pkl` is preprocessed consistently with ours.
> * **Hardware differences**: Different GPU architectures may introduce minor numerical differences.
> 
> Could you share the specific metrics you obtained compared to the expected values? This would help us better diagnose the issue.
> 
> **Regarding Question 2 (Homogeneous features not in final_rep)**
> 
> Thank you for catching this! This was an oversight when we merged the ablation study versions into the final codebase. The homogeneous features should indeed contribute to the final representation.
> 
> We have now fixed this issue. The updated fusion strategy is:
> 
> # Homogeneity branch: average pooling over time dimension
> c_l_avg = c_l.mean(dim=2)
> c_a_avg = c_a.mean(dim=2)
> c_v_avg = c_v.mean(dim=2)
> fusion_rep_homo = torch.cat([c_l_avg, c_a_avg, c_v_avg], dim=1)  # [N, 3*d_l]
> 
> # Fuse heterogeneous paths and concatenate with homogeneous
> fusion_rep_hete = fusion_rep_trans + fusion_rep_cma  # [N, 3*d_l]
> final_rep = torch.cat([fusion_rep_hete, fusion_rep_homo], dim=1)  # [N, 6*d_l]
> 
> Thank you again for your feedback! Please let me know if this addresses all your concerns.
Hello, author. Could you help me identify the problem
[no_bert.txt](https://github.com/user-attachments/files/25088519/no_bert.txt)
[use_bert.txt](https://github.com/user-attachments/files/25088520/use_bert.txt)
The above file is the result of changing "use_bert": true in config, and the dataset used is MOSI. Other configurations are set to default
Dataset usage: https://drive.google.com/drive/folders/1A2S4pqCHryGmiqnNSPLv7rEg63WvjCSk

**Comment by Raymond-Qiancx** [comment]
Hi, thank you for your careful observation.

After re-checking the released implementation and our internal experimental versions, we found that some code paths from different development stages were not fully synchronized during version management. This may have caused confusion around how the homogeneous representations are used in the final prediction branch.

We have now systematically reviewed and corrected the implementation. In the updated version, the final representation explicitly incorporates both the heterogeneous/modality-specific branch and the homogeneous/modality-common branch:

```python
# heterogeneous / modality-specific representations
fusion_rep_hete = torch.cat([h_l, h_a, h_v], dim=1)

# homogeneous / modality-common representations
c_l_avg = c_l.mean(dim=2)
c_a_avg = c_a.mean(dim=2)
c_v_avg = c_v.mean(dim=2)
fusion_rep_homo = torch.cat([c_l_avg, c_a_avg, c_v_avg], dim=1)

# final prediction representation
final_rep = torch.cat([fusion_rep_hete, fusion_rep_homo], dim=1)
```

The updated implementation is now aligned with the intended framework design, where both modality-unique and modality-common representations contribute to the final prediction.

Thanks again for pointing this out. This helped us improve the consistency of the released code.

**Comment by Raymond-Qiancx** [comment]
Closing this for now since the questions have been addressed above and there has been no further follow-up. Please feel free to reopen with detailed configs and logs if the issue persists.
